php/libraries/DefaultFilesUploadHandler.class.inc

```diff
@@ -0,0 +1,367 @@
+    }
+
+    /**
+     * Accessor for field $_movedFiles.
+     *
+     * @return array the final location of all files uploaded on the
+     *               server. Each key is the name of a LorisFirm field
+     *               used to upload a file and its corresponding value is the final
+     *               location for the uploaded file on the server.
+     */
+    public function getMovedFiles() : array
+    {
+        return $this->_movedFiles;
+    }
+
+    /**
+     * Accessor for field $_dirPermissions.
+     *
+     * @return int value for field $_dirPermissions.
+     */
+    public function getFiles() : array
+    {
+        if (is_null($this->_files)) {
+            $this->_files = array();
+            foreach ($this->_lorisForm->getElements() as $element) {
+                if ($element instanceof LorisFormFileElement) {
+                    $this->_files[$element->getName()] = $element->getValue();
+                }
+            }
+        }
+
+        return $this->_files;
+    }
+
+    /**
+     * Determines if the set of uploaded files is valid or not, updating the $_errors
+     * array with the errors found. This method will look at all the fields of the
+     * LorisForm associated to this handler and verify that the upload error code
+     * found in the super global $_FILES for that field is UPLOAD_ERR_OK
+     * (see documentation for $_FILES). If no file name was entered for a given field
+     * in the LorisForm, then the upload is considered invalid and the error for that
+     * field will be set to false. This method assumes that all the fields used for
+     * file uploads have a value (i.e. no "optional" upload is allowed). This
+     * behavior can be overridden bu subclasses should they need to.
+     *
+     * @return bool true if the uploaded file set is valid, false otherwise.
+     */
+    public function isValid() : bool
+    {
+        $valid = true;
+        foreach ($this->getFiles() as $field => $file) {
+            if ($file === false || $file['error'] != UPLOAD_ERR_OK) {
+                $this->addError(
+                    $field,
+                    $file === false ? false : $file['error']
+                );
+                $valid = false;
+            }
+        }
+
+        return $valid;
+    }
+
+    /**
+     * Main method for this handler. Validates the uploaded files (i.e. make sure
+     * that they are are present on the server) and moves them to their final
+     * locations.
+     *
+     * @return nothing.
+     */
+    public function processAll()
+    {
+        try {
+            if ($this->isValid()) {
+                foreach ($this->getFiles() as $field => $file) {
+                    if ($file !== false) {
+                        $newFile = $this->getNewLocation($file['tmp_name']);
+                        $newDir  = $newFile->getPath();
+
+                        // Check to see if directory exists
+                        if (!file_exists($newDir)) {
```

Comment:
This whole block that check if the directory exists and create it if not should be encapsulated in a function so this `processAll()` would look less convoluted.

Reply:
Good idea. I'll refactor.